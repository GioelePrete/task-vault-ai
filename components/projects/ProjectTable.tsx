import React from "react";
import ProjectTableRow from "@/components/projects/ProjectTableRow";

// Define the props interface for the component
interface Props {
    projects: ProjectData[];
    setProjectModalMode: (projectModalMode: "add" | "edit") => void;
    openProjectModal: () => void
    deleteProject: (id: number) => void;
}

const ProjectTable = ({projects, setProjectModalMode, openProjectModal, deleteProject}: Props) => {
    return (
        <table className="table-auto w-full">
            {/* Table Header */}
            <thead className="bg-gray-300 text-left h-16">
            <tr>
                <th className="px-6 py-3">ID</th>
                <th className="px-6 py-3">Owner</th>
                <th className="px-6 py-3">Name</th>
                <th className="px-6 py-3">Description</th>
                <th className="px-6 py-3">Status</th>
                <th className="px-6 py-3">Progress</th>
                <th className="px-6 py-3">Priority</th>
                <th className="px-6 py-3">Actions</th>
            </tr>
            </thead>
            {/* Table Body */}
            <tbody className="bg-white divide-y divide-gray-200">
            {/* Iterate through each project and render a ProjectTableRow */}
            {projects.map((project: ProjectData, index: number) => (
                <ProjectTableRow
                    key={project.id} // Make sure to provide a unique key for React's reconciliation
                    project={project}
                    index={index}
                    setProjectModalMode={setProjectModalMode}
                    openProjectModal={openProjectModal}
                    deleteProject={deleteProject}
                />
            ))}
            </tbody>
        </table>
    );
};

export default ProjectTable;
