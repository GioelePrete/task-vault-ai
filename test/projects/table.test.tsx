import React from "react";
import {render, screen} from "@testing-library/react";
import ProjectTable from "@/components/projects/ProjectTable";

// Mock the ProjectTableRow component
jest.mock("@/components/projects/ProjectTableRow", () => {
    return function MockProjectTableRow(props) {
        return (
            <tr data-testid="project-row">
                <td>{props.project.id}</td>
                <td>{props.project.name}</td>
                <td>{props.project.description}</td>
                <td>{props.project.status}</td>
                <td>{props.project.progress}</td>
                <td>{props.project.status}</td>
            </tr>
        );
    };
});

describe("ProjectTable", () => {
    const mockProjects = [
        {
            id: 1,
            name: "Project A",
            description: "New description project A",
            status: "pending",
            progress: 20,
            priority: "medium"
        },
        {
            id: 2,
            name: "Project B",
            description: "New description project B",
            status: "completed",
            progress: 100,
            priority: "low"
        },
    ];

    it("renders table rows for each project", () => {
        render(
            <ProjectTable
                projects={mockProjects}
                setProjectModalMode={() => {
                }}
                openProjectModal={() => {
                }}
                deleteProject={() => {
                }}
            />
        );

        const projectRows = screen.getAllByTestId("project-row");
        expect(projectRows).toHaveLength(mockProjects.length);
    });

    it("renders correct project ID in the table row", () => {
        render(
            <ProjectTable
                projects={mockProjects}
                setProjectModalMode={() => {
                }}
                openProjectModal={() => {
                }}
                deleteProject={() => {
                }}
            />
        );

        const projectIDCells = screen.getAllByText(/^1$|^2$/); // Use the corrected regular expression
        expect(projectIDCells).toHaveLength(mockProjects.length); // You should expect 2 cells for each project
    });

    it("matches snapshot", () => {
        const {container} = render(
            <ProjectTable
                projects={mockProjects}
                setProjectModalMode={() => {
                }}
                openProjectModal={() => {
                }}
                deleteProject={() => {
                }}
            />
        );

        expect(container).toMatchSnapshot();
    });
});
