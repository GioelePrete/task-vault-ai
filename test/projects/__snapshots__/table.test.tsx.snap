// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectTable matches snapshot 1`] = `
<div>
  <table
    class="table-auto w-full"
  >
    <thead
      class="bg-gray-300 text-left h-16"
    >
      <tr>
        <th
          class="px-6 py-3"
        >
          ID
        </th>
        <th
          class="px-6 py-3"
        >
          Owner
        </th>
        <th
          class="px-6 py-3"
        >
          Name
        </th>
        <th
          class="px-6 py-3"
        >
          Description
        </th>
        <th
          class="px-6 py-3"
        >
          Status
        </th>
        <th
          class="px-6 py-3"
        >
          Progress
        </th>
        <th
          class="px-6 py-3"
        >
          Priority
        </th>
        <th
          class="px-6 py-3"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      class="bg-white divide-y divide-gray-200"
    >
      <tr
        data-testid="project-row"
      >
        <td>
          1
        </td>
        <td>
          Project A
        </td>
        <td>
          New description project A
        </td>
        <td>
          pending
        </td>
        <td>
          20
        </td>
        <td>
          pending
        </td>
      </tr>
      <tr
        data-testid="project-row"
      >
        <td>
          2
        </td>
        <td>
          Project B
        </td>
        <td>
          New description project B
        </td>
        <td>
          completed
        </td>
        <td>
          100
        </td>
        <td>
          completed
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
